reqrespactivity/requester.go

```diff
@@ -0,0 +1,130 @@
+package reqrespactivity
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/pborman/uuid"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/worker"
+)
+
+// Requester can request uppercasing of strings and should be closed after use.
+type Requester struct {
+	options   RequesterOptions
+	taskQueue string
+	// Channels need buffer of 1 because they are sent to in non-blocking fashion
+	pendingRequests     map[string]chan<- *Response
+	pendingRequestsLock sync.RWMutex
+}
+
+// RequesterOptions are options for NewRequester.
+type RequesterOptions struct {
+	// Client to the Temporal server. Required. Not closed on Requester.Close.
+	Client client.Client
+	// ID of the workflow listening for signals to uppercase. Required.
+	TargetWorkflowID string
+
+	// Visible for testing.
+	ExistingWorker interface {
+		RegisterActivity(interface{})
+		Start() error
+		Stop()
+	}
+}
+
+// NewRequester creates a new Requester for the given options.
+func NewRequester(options RequesterOptions) (*Requester, error) {
+	if options.Client == nil {
+		return nil, fmt.Errorf("client required")
+	} else if options.TargetWorkflowID == "" {
+		return nil, fmt.Errorf("target workflow required")
+	}
+
+	// Create requester
+	r := &Requester{
+		options:         options,
+		taskQueue:       "requester-" + uuid.New(),
+		pendingRequests: map[string]chan<- *Response{},
+	}
+	if r.options.ExistingWorker == nil {
+		r.options.ExistingWorker = worker.New(options.Client, r.taskQueue, worker.Options{})
+	}
+
+	// Start worker
+	r.options.ExistingWorker.RegisterActivity(r.responseActivity)
```

Comment:
As far as I can tell would mean that the worker will poll on an increasing number of task queues. It's not clear to me if this is factored this way simply because of the constraints of making a sample or if this is a suggestion for a useable pattern. I appreciate the need for separated task queues so that the response lands in the same process that it was requested in but this does feel a bit odd.

Reply:
Yes, there is a 1:1 correlation with workers/taskqueues and requesters. If you have 50 places sending request/response, you have 50 workers polling on separate task queues. For the most part an open long poll on each is mostly just a harmless HTTP connection (especially when I complete https://github.com/temporalio/sdk-go/issues/636 here in just a bit). Is it really any different than 50 client-side `ExecuteWorkflow` requests that all long poll on `Get` when awaiting response?

I definitely encourage long-term use of this "Requester", not per call. In fact, I would suggest only one per process.

I am open to alternatives of course.

Reply:
The difference is that the pollers would stop long polling once they've got their answer. I may have missed it in the code but it looked to me that the activity is registered and never de-registered (is that even a thing?) which means it could add up to a pretty large number of pollers who will never see any traffic again (depending on request count, process lifetime etc etc).

Reply:
Yes, I think there's some confusion here.

Say this requester should be reused for the all requests for the lifetime of the process. That means only one worker (so only one poller) and one activity registered to it. There is not a response activity per request. There is no need to unregister an activity, the activity is 1:1 with the worker.

So basically, if used properly, there is only one of these per process.

Reply:
> The difference is that the pollers would stop long polling once they've got their answer.

I do not believe this is true. Activity pollers will continuously poll for the lifetime of the worker.

Reply:
Indeed I was confused, I mis-read 'requester-uuid' as 'request-uuid' and had inferred it was one task queue per request/response pair, not per process. Makes sense. My statement about pollers stopping was referring to your mention of people calling StartWorkflow and then polling for the result. I believe they would generally stop polling once they got the result.